File: Poker.hs
-- Name: Farhan Sami -- RSN: 500892031
-- Name: Sameer Naumani -- RSN: 500703778

module Poker
(deal) where

import Data.List

-- deal function: produces a list of strings from winner function return
deal cards = do
    let win = winner cards

    let one = head win
    let two = head (tail win)
    let three = head (tail (tail win))
    let four = head (tail (tail (tail win)))
    let five = last win

    let result = suit one : suit two : suit three : suit four : suit five :[]

    result

-- executes all conditions and decides winner using the Poker rule logic
winner cards = do
    let hand1 = sort (evens cards)
    let hand2 = sort (odds cards)

    let royalFlush1 = royalFlush hand1
    let royalFlush2 = royalFlush hand2

    let straightFlush1 = straightFlush hand1
    let straightFlush2 = straightFlush hand2

    let fourKind1 = fourKind hand1
    let fourKind2 = fourKind hand2

    let fullHouse1 = fullHouse hand1
    let fullHouse2 = fullHouse hand2

    let flush1 = flush hand1
    let flush2 = flush hand2

    let straight1 = straight hand1
    let straight2 = straight hand2

    let threeKind1 = threeKind hand1
    let threeKind2 = threeKind hand2

    let twoPairs1 = twoPairs hand1
    let twoPairs2 = twoPairs hand2

    let pairs1 = pairs hand1
    let pairs2 = pairs hand2

    if royalFlush1 || royalFlush2
    then if royalFlush1 && royalFlush2
        then highCard hand1 hand2
        else if royalFlush1
            then hand1
            else hand2

    else if straightFlush1 || straightFlush2
        then  if straightFlush1 && straightFlush2
            then highCard hand1 hand2
            else if straightFlush1
                then hand1
                else hand2

        else if fourKind1 || fourKind2
                then if fourKind1 && fourKind2
                    then highCard hand1 hand2
                    else if fourKind1
                        then hand1
                        else hand2

            else if fullHouse1 || fullHouse2
                    then if fullHouse1 && fullHouse2
                        then highCard hand1 hand2
                        else if fullHouse1
                            then hand1
                            else hand2

                else if flush1 || flush2
                        then if flush1 && flush2
                            then highCard hand1 hand2
                            else if flush1
                                then hand1
                                else hand2

                    else if straight1 || straight2
                            then if straight1 && straight2
                                then highCard hand1 hand2
                                else if straight1
                                    then hand1
                                    else hand2

                        else if threeKind1 || threeKind2
                                then if threeKind1 && threeKind2
                                    then highCard hand1 hand2
                                    else if threeKind1
                                        then hand1
                                        else hand2

                            else if twoPairs1 || twoPairs2
                                    then if twoPairs1 && twoPairs2
                                        then highCard hand1 hand2
                                        else if twoPairs1
                                            then hand1
                                            else hand2

                                else if pairs1 || pairs2
                                        then if pairs1 && pairs2
                                            then highCard hand1 hand2
                                            else if pairs1
                                                then hand1
                                                else hand2 

                                    else highCard hand1 hand2

-- return true if hand is a royal flush
royalFlush :: [Integer] -> Bool
royalFlush hand
    | hand == [1,10,11,12,13] = True
    | hand == [14,23,24,25,26] = True
    | hand == [27,36,37,38,39] = True
    | hand == [40,49,50,51,52] = True
    | otherwise = False

-- return true if hand is a straight flush
straightFlush hand = flush hand && straight hand

-- return true if hand is a four kind
fourKind hand = do
    let kind = rank hand
    let one = head kind
    let two = head (tail kind)
    let three = head (tail (tail kind))
    let four = head (tail (tail (tail kind)))
    let five = last kind
    if (one == two && two == three && three == four) || (five == two && two == three && three == four)
    then True
    else False    

-- return true if hand is a full house
fullHouse hand = do
    let kind = rank hand
    let one = head kind
    let two = head (tail kind)
    let three = head (tail (tail kind))
    let four = head (tail (tail (tail kind)))
    let five = last kind
    if (one == two && three == four && three == five) || (one == two && one == three && four == five)
    then True
    else False

-- return true if hand is a flush
flush :: [Integer] -> Bool
flush hand
    | subList hand [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13] == True = True
    | subList hand [14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26] == True = True
    | subList hand [27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39] == True = True
    | subList hand [40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52] == True = True
    | otherwise = False

-- return true if hand is a straight
straight hand = do
    let kind = rank hand
    let first = head kind
    let last = first + 4
    let straight = [x | x <- [first..last]]
    if kind == straight
    then True
    else False

-- return true if hand is a three kind
threeKind hand = do
    let kind = rank hand
    let middle = head (tail (tail kind))
    if subList [middle, middle, middle] kind
    then True
    else False 

-- return true if hand is a two pair
twoPairs hand = do
    let kind = rank hand
    let one = head kind
    let two = head (tail kind)
    let three = head (tail (tail kind))
    let four = head (tail (tail (tail kind)))
    let five = last kind
    if (count one kind == 2 && ((count two kind == 2 && one /= two) || count three kind == 2 || count four kind == 2 || count five kind == 2) || count two kind == 2 && ((count one kind == 2 && one /= two) || (count three kind == 2 && two /= three) || count four kind == 2 || count five kind == 2) || count three kind == 2 && ((count two kind == 2 && two /= three) || count one kind == 2 || (count four kind == 2 && three /= four) || count five kind == 2) || count four kind == 2 && (count two kind == 2 || (count three kind == 2 && three /= four) || count one kind == 2 || (count five kind == 2 && four /= five)) || count five kind == 2 && (count two kind == 2 || count three kind == 2 || (count four kind == 2 && four /= five) || count one kind == 2))
    then True
    else False

-- return true if hand is a pair
pairs hand = do
    let kind = rank hand
    let one = head kind
    let three = head (tail (tail kind))
    let five = last kind
    if count one kind == 2 || count three kind == 2 || count five kind == 2
    then True
    else False

-- returns the higher of two hands
highCard hand1 hand2 = do
    let kind1 = rank hand1
    let kind2 = rank hand2
    if hand1 == hand2
    then best hand1 hand2
    else if hand1 > hand2
        then hand1
        else hand2

-- returns the hand with cards from higher suits
best hand1 hand2 = do
    if sum hand1 > sum hand2
    then hand1
    else hand2

-- converts all suits to single rank
rank hand = sort ([x `mod` 13 | x <- hand])

-- converts the number to rank and returns it as string
rankstr num = show (num `mod` 13)

-- returns the letter for suit that the card belongs to
suit :: Integer -> String
suit card
    | card == 11 = "Jack C" 
    | card == 12 = "Queen C"
    | card == 13 = "King C"
    | card == 24 = "Jack D"  
    | card == 25 = "Queen D" 
    | card == 26 = "King D" 
    | card == 37 = "Jack H"  
    | card == 38 = "Queen H" 
    | card == 39 = "King H" 
    | card == 50 = "Jack S"  
    | card == 51 = "Queen S" 
    | card == 52 = "King S" 
    | card > 0 && card < 14 = rankstr card ++ "C"
    | card > 13 && card < 27 = rankstr card ++ "D"
    | card > 26 && card < 40 = rankstr card ++ "H"
    | card > 39 && card < 53 = rankstr card ++ "S"
    | otherwise = "invalid card"

-- checks if a list is sub list of another function
subList :: Eq a => [a] -> [a] -> Bool
subList [] [] = True
subList _ []    = False
subList [] _    = True
subList (x:xs) (y:ys) 
    | x == y    = subList xs ys   
    | otherwise = subList (x:xs) ys

-- counts the number of occurences of the element in list
count   :: Eq a => a -> [a] -> Int
count x =  length . filter (==x)

-- splits list into lists of even and odd indices
evens (x:xs) = x:odds xs
evens _ = []
odds (_:xs) = evens xs
odds _ = []

File: poker.py
import random

def winner(deal):
    hand = poker(deal)
    cards = []

    for x in hand:
        if x > 0 and x < 14:
            cards.append(str(x)+"C")
        if x > 13 and x < 27:
            cards.append(str(1+x%13)+"D")
        if x > 26 and x < 40:
            cards.append(str(1+x%13)+"H")
        if x > 39 and x < 53:
            cards.append(str(1+x%13)+"S")
    print(cards)

def poker(deal):
    hand1 = [deal[i] for i in range(0, 10, 2)]
    hand2 = [deal[i] for i in range(1, 10, 2)]

    print("Deal:", deal)
    print("Hand 1:", hand1)
    print("Hand 2:", hand2)

    royalFlush1 = royalFlush(hand1)
    royalFlush2 = royalFlush(hand2)

    straightFlush1 = straightFlush(hand1)
    straightFlush2 = straightFlush(hand2)

    fourKind1 = fourKind(hand1)
    fourKind2 = fourKind(hand2)

    fullHouse1 = fullHouse(hand1)
    fullHouse2 = fullHouse(hand2)

    flush1 = flush(hand1)
    flush2 = flush(hand2)

    straight1 = straight(hand1)
    straight2 = straight(hand2)

    threeKind1 = threeKind(hand1)
    threeKind2 = threeKind(hand2)

    twoPairs1 = twoPairs(hand1)
    twoPairs2 = twoPairs(hand2)

    pairs1 = pairs(hand1)
    pairs2 = pairs(hand2)

    if royalFlush1 or royalFlush2:
        if royalFlush1 and royalFlush2:return highCard(hand1, hand2)
        if royalFlush1:return hand1
        else:return hand2

    elif straightFlush1 or straightFlush2:
        if straightFlush1 and straightFlush2:return highCard(hand1, hand2)
        if straightFlush1:return hand1
        else:return hand2

    elif fourKind1 or fourKind2:
        if fourKind1 and fourKind2:return highCard(hand1, hand2)
        if fourKind1:return hand1
        else:return hand2

    elif fullHouse1 or fullHouse2:
        if fullHouse1 and fullHouse2:return highCard(hand1, hand2)
        if fullHouse1:return hand1
        else:return hand2

    elif flush1 or flush2:
        if flush1 and flush2:return highCard(hand1, hand2)
        if flush1:return hand1
        else:return hand2

    elif straight1 or straight2:
        if straight1 and straight2:return highCard(hand1, hand2)
        if straight1:return hand1
        else:return hand2

    elif threeKind1 or threeKind2:
        if threeKind1 and threeKind2:return highCard(hand1, hand2)
        if threeKind1:return hand1
        else:return hand2

    elif twoPairs1 or twoPairs2:
        if twoPairs1 and twoPairs2:return highCard(hand1, hand2)
        if twoPairs1:return hand1
        else:return hand2

    elif pairs1 or pairs2:
        if pairs1 and pairs2:return highCard(hand1, hand2)
        if pairs1:return hand1
        else:return hand2

    else:
        return highCard(hand1, hand2)

def sortCards(hand1, hand2):
    hand1.sort()
    hand2.sort()

def royalFlush(hand):
    royals = [[1, 10, 11, 12, 13], [i+13 for i in [1, 10, 11, 12, 13]], [i+26 for i in [1, 10, 11, 12, 13]], [i+39 for i in [1, 10, 11, 12, 13]]]
    if hand in royals: return True
    return False

def straightFlush(hand):
    for x in range(0, 4):
        if (hand[x+1] - hand[x]) != 1:
            return False
    return True

def fourKind(hand):
    kind = rank(hand)
    if kind.count(kind[0]) == 4 or kind.count((kind[-1])) == 4: return True
    return False

def fullHouse(hand):
    kind = rank(hand)
    if (kind.count(kind[0]) == 2 and kind.count((kind[-1])) == 3) or (kind.count(kind[0]) == 3 and kind.count((kind[-1])) == 2): return True
    return False

def flush(hand):
    suits = [[i for i in range(1, 14)], [i+13 for i in range(1, 14)], [i+26 for i in range(1, 14)], [i+39 for i in range(1, 14)]]
    for x in suits:
        if set(hand).issubset(set(x)):
            return True
    else:return False

def straight(hand):
    kind = rank(hand)
    for i in range(0, 4):
        if kind[i+1] - kind[i] != -1:return False
    else:return True

def threeKind(hand):
    kind = rank(hand)
    if kind.count(kind[2]) == 3:return True
    return False

def twoPairs(hand):
    kind = rank(hand)
    ret = False
    if kind.count(kind[0]) == 2:
        rem = kind[0]
        kind.remove(rem)
        kind.remove(rem)
        if kind.count(kind[0])== 2 or kind.count(kind[2]) == 2:ret = True
    elif kind.count(kind[2]) == 2:
        rem = kind[2]
        kind.remove(rem)
        kind.remove(rem)
        if kind.count(kind[0])== 2 or kind.count(kind[2]) == 2:ret = True
    elif kind.count(kind[4]) == 2:
        rem = kind[4]
        kind.remove(rem)
        kind.remove(rem)
        if kind.count(kind[0])== 2 or kind.count(kind[2]) == 2:ret = True
    return ret

def pairs(hand):
    kind = rank(hand)
    if kind.count(kind[0]) == 2 or kind.count(kind[2]) == 2 or kind.count(kind[4]) == 2:return True
    return False

def highCard(hand1, hand2):
    kind1 = rank(hand1)
    kind2 = rank(hand2)
    for x in range(0, 5):
        if kind1[x] != kind2[x]:
            if kind1[x] > kind2[x]:
                return hand1
            else:
                return hand2
    else:
        return best(hand1, hand2)

def best(hand1, hand2):
    if sum(hand1)>sum(hand2):return hand1
    else:return hand2

def rank(hand):
    kind = [i%13 for i in hand]
    kind.sort(reverse=True)
    return kind


for x in range(1):
    print("Test", x)
    winner([random.randrange(1, 53) for i in range(1, 11)])